src/core/connection.c

```diff
@@ -3076,6 +3076,35 @@ QuicConnQueueUnreachable(
     }
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+QuicConnQueueRouteCompletion(
+    _In_ QUIC_CONNECTION* Connection,
+    _When_(Succeeded == FALSE, _In_opt_)
+    _When_(Succeeded == TRUE, _In_)
+        const uint8_t* PhysicalAddress,
+    _In_ BOOLEAN Succeeded
+    )
+{
+    QUIC_OPERATION* ConnOper =
+        QuicOperationAlloc(Connection->Worker, QUIC_OPER_TYPE_ROUTE_COMPLETION);
+    if (ConnOper != NULL) {
+        ConnOper->ROUTE.Succeeded = Succeeded;
+        if (Succeeded) {
+            memcpy(ConnOper->ROUTE.PhysicalAddress, PhysicalAddress, sizeof(ConnOper->ROUTE.PhysicalAddress));
+        }
+        QuicConnQueueOper(Connection, ConnOper);
+    } else {
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "Route completion operation",
+            0);
+    }
+
+    QuicConnRelease(Connection, QUIC_CONN_REF_ROUTE);
```

Comment:
This releases the extra ref taken in QuicSendQueueFlush if ResolveRoute returns pending.